compiler/src/dotty/tools/dotc/sbt/ExtractAPI.scala

```diff
@@ -620,27 +620,130 @@ private class ExtractAPICollector(using Context) extends ThunkHolder {
+      h
+    end typeHash
+
+    def constantHash(c: Constant, initHash: Int): Int =
+      var h = MurmurHash3.mix(initHash, c.tag)
+      c.tag match
+        case NullTag =>
+          // No value to hash, the tag is enough.
+        case ClazzTag =>
+          h = typeHash(c.typeValue, h)
+        case _ =>
+          h = MurmurHash3.mix(h, c.value.hashCode)
+      h
+    end constantHash
+
+    /**An inline method that calls another inline method will eventually inline the call
+     * at a non-inline callsite, in this case if the implementation of the nested call
+     * changes, then the callsite will have a different API, we should hash the definition
+     */
+    def inlineDefHash(ref: Symbol, initHash: Int): Int =
+      var h = initHash
+
+      def paramssHash(paramss: List[List[Symbol]], initHash: Int): Int = paramss match
+        case Nil :: paramss1 =>
+          paramssHash(paramss1, MurmurHash3.mix(initHash, EmptyParamHash))
+        case params :: paramss1 =>
+          var h = initHash
+          val paramsIt = params.iterator
+          while paramsIt.hasNext do
+            val param = paramsIt.next
+            h = nameHash(param.name, h)
+            h = typeHash(param.info, h)
+            if param.is(Inline) then
+              h = MurmurHash3.mix(h, InlineParamHash) // inline would change the generated code
+            if param.isType then
+              val variance = param.paramVarianceSign
+              if variance != 0 then
+                h = MurmurHash3.mix(h, variance)
+          paramssHash(paramss1, h)
+        case Nil =>
+          initHash
+      end paramssHash
+
+      h = paramssHash(ref.paramSymss, h)
+      h = typeHash(ref.info.finalResultType, h)
+      positionedHash(Inliner.bodyToInline(ref), h)
+    end inlineDefHash
+
+    def err(what: String, elem: Any, pos: Positioned, initHash: Int): Int =
+      internalError(i"Don't know how to produce a stable hash for $what", pos.sourcePos)
+      MurmurHash3.mix(initHash, elem.toString.hashCode)
 
     def positionedHash(p: ast.Positioned, initHash: Int): Int =
+      var h = initHash
+
       p match
         case p: WithLazyField[?] =>
           p.forceIfLazy
         case _ =>
+
+      p match
+        case ref: RefTree @unchecked =>
+          val sym = ref.symbol
+          if sym.is(Inline, butNot = Param) && !seenInlines.contains(sym) then
+            seenInlines += sym // dont re-enter hashing this ref
+            if sym.is(Method) then // inline def
+              if Inliner.hasBodyToInline(sym) then
+                h = inlineDefHash(sym, h)
+              else
+                h = err(i"inline method reference `${ref.name}`", ref.name, ref, h)
+            else // inline val
+              // before inlining, inline val can be any expression/type, after inlining it should be constant
+              val tp = sym.info.widenTermRefExpr.dealias.normalized // copied from `constToLiteral`
+              h = typeHash(tp, h)
```

Comment:
again, should we go through `defTree` to get the rhs? Perhaps we should also attach a `BodyAnnotation`?

Reply:
Going through defTree should be fine, it should always be available on things marked inline. The only reason we also have BodyAnnotation for inline defs is that the method body has to be transformed to add private accessors etc.

Reply:
it seems defTree is not available for `scala.compiletime.error` e.g. when compiling `scala3-library-bootstrappedJS`

Reply:
`scala.compiletime.error` does still have `bodyToInline` however `???: Nothing` is not very useful, perhaps I just escape that package with some constant hash

Reply:
It would be good to figure out how come it doesn't have a defTree though, when compiling scala3-library-bootstrappedJS scala.compiletime.error should be among the sources being compiled so it should get a defTree in PostTyper: https://github.com/lampepfl/dotty/blob/7627583448ddde71c3f19c8ec962f3b16f917652/compiler/src/dotty/tools/dotc/transform/PostTyper.scala#L142

Reply:
I did a clean compile and it seems this was not an issue anymore 😂

Reply:
Ah but in that case maybe defTree isn't available under separate compilation which would still be a problem (since it could lead to different hashes depending on what files are jointly or separately compiled), but the unpickler calsl setDefTree so I don't know why that would happen: https://github.com/lampepfl/dotty/blob/7627583448ddde71c3f19c8ec962f3b16f917652/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala#L899

Reply:
I'm not really sure how to replicate this if it is possible

Reply:
Add an assert that the defTree is present, then compile scala3-library-bootstrapped, edit a file that contains a macro that calls scala.compiletime.error (in a file other than where `error` is defined) and see if it triggers? Even better would be to print the hash of the macro under separate and joined compilation to check that it didn't change.

Reply:
So to recreate this issue - I ran the task `sjsSandbox/run`, then I added a new file `library/src/scala/compiletime/package2.scala` with contents:
```scala
package scala.compiletime

inline def foo: Nothing =
  error("who knows...")
```
and now I get the same error:
```scala
[error] -- Error: /Users/jamie/workspace/dotty/library/src/scala/compiletime/package2.scala:4:2 
[error] 4 |  error("who knows...")
[error]   |  ^^^^^
[error]   |Internal error in the incremental compiler while compiling /Users/jamie/workspace/dotty/library/src/scala/compiletime/package2.scala: Don't know how to produce a stable hash for inline method reference `error`
```
implying that `defTree` was the empty tree

Reply:
I did a further check to print calls to `defTree_=` when running `sjsSandbox/run`, and it is only called for the `inline msg: String` parameter of `compiletime.package.error` and nothing else in `compiletime.package`

Reply:
this can be simplified further by running the `scala3-library-bootstrapped/compile` task, modifying `scala.compiletime.error`, and then re-running `scala3-library-bootstrapped/compile` - it appears that there are two symbols for `scala.compiletime.error` - one that gets assigned a def tree, and one that does not

Reply:
The cause of the error is that the `compiletime.error` method is not receiving a def tree when recompiling a method that depends on it. To reproduce do the following:

- run `scala3-library-bootstrapped/compile` task
- modify `scala.compiletime.error` somehow, e.g. change its rhs to `{???;???}`
- re-run `scala3-library-bootstrapped/compile`
 
What happens on the third step is as follows:
- Zinc first compiles `library/src/scala/compiletime/package.scala` with the non-bootstrapped compiler,
- Zinc then compiles the file again with the bootstrapped compiler.
- Zinc then begins a new compilation run on `library/src/scala/compiletime/testing/package.scala` aka a different file that depends on the first one
  - the compiler tries to pickle the reference to `scala.compiletime.error` in the body of `scala.compiletime.testing.typeChecks`, but it does not have a def tree

Reply:
not sure how to recreate this in the sbt scripted tests

Reply:
To fix this I have gone back to using the inline body annotation. The reason being that symbols from the sourcepath are type checked only with `-Yretain-trees` and also after all other phases have run. The inline body annotation however is installed with the ability to force type checking of the inline definition, even if from the sourcepath

Reply:
this problem was fixed by #14050 , and tested in https://github.com/lampepfl/dotty/pull/12931/commits/1c560903671b3c9c630e551d1ef3386a154c3100